extra/src/main/java/aima/extra/probability/BigDecimalProbabilityNumber.java

```diff
@@ -0,0 +1,381 @@
+	 * BigDecimal.valueOf() builds from a double type based on its canonical
+	 * representation. Precision of double type is at max 15.95 (~15).
+	 * 
+	 * @param value
+	 *            to be assigned.
+	 */
+	public BigDecimalProbabilityNumber(double value) {
+		if (value < 0 || value > 1) {
+			throw new IllegalArgumentException();
+		}
+		this.value = new BigDecimal(Double.toString(value), precisionMathContext);
+	}
+
+	/**
+	 * Construct a BigDecimalProbabilityNumber from BigDecimal type.
+	 * 
+	 * @param value
+	 *            to be assigned.
+	 */
+	public BigDecimalProbabilityNumber(BigDecimal value) {
+		if (null == value || value.compareTo(BIG_DECIMAL_ZERO) == -1 || value.compareTo(BIG_DECIMAL_ONE) == 1) {
+			throw new IllegalArgumentException();
+		}
+		this.value = new BigDecimal(value.toString(), precisionMathContext);
+	}
+
+	/**
+	 * Construct a BigDecimalProbabilityNumber from BigDecimal type.
+	 * 
+	 * @param value
+	 *            to be assigned.
+	 * 
+	 * @param mc
+	 *            MathContext to be associated with value.
+	 */
+	public BigDecimalProbabilityNumber(BigDecimal value, MathContext mc) {
+		precisionMathContext = mc;
+		setConstants();
+		if (value.compareTo(BIG_DECIMAL_ZERO) == -1 || value.compareTo(BIG_DECIMAL_ONE) == 1) {
+			throw new IllegalArgumentException();
+		}
+		this.value = new BigDecimal(value.toString(), precisionMathContext);
+	}
+
+	// Private methods
+
+	/**
+	 * Initialize constants BIG_DECIMAL_ZERO and BIG_DECIMAL_ONE according to
+	 * precisionMathContext that is defined.
+	 */
+	private void setConstants() {
+		BIG_DECIMAL_ZERO = new BigDecimal(0, precisionMathContext);
+		BIG_DECIMAL_ONE = new BigDecimal(1, precisionMathContext);
+	}
+
+	/**
+	 * Covert other implementations of the ProbabilityNumber interface to a
+	 * BigDecimalProbabilityNumber.
+	 * 
+	 * @param that
+	 *            BigDecimalProbabilityNumber.
+	 */
+	private BigDecimalProbabilityNumber toInternalType(ProbabilityNumber that) {
+		if (that instanceof BigDecimalProbabilityNumber) {
+			return (BigDecimalProbabilityNumber) that;
+		}
+		return new BigDecimalProbabilityNumber(that.getValue(), that.getMathContext());
+	}
+
+	/**
+	 * Compare two BigDecimal values upto minimum precision.
+	 * 
+	 * @param first
+	 *            of BigDecimal type.
+	 * @param second
+	 *            of BigDecimal type.
+	 * 
+	 * @return 1 if first > second, 0 if first == second, -1 if first < second.
+	 */
+	private int compareBigDecimal(BigDecimal first, MathContext mc1, BigDecimal second, MathContext mc2) {
+		if (mc1.getPrecision() == mc2.getPrecision()) {
+			return first.compareTo(second);
+		}
+		int minPrecision = getMinPrecision(mc1.getPrecision(), mc2.getPrecision());
+		BigDecimal valueA = new BigDecimal(first.toString(), new MathContext(minPrecision, mc1.getRoundingMode()));
+		BigDecimal valueB = new BigDecimal(second.toString(), new MathContext(minPrecision, mc2.getRoundingMode()));
+		return valueA.compareTo(valueB);
```

Comment:
Isn't the above unnecessarily creating two intermediate instances (expensively as well using toString) when at most you should only every require 1 and in some cases none if they start with the same precision.